code/modules/SCP/scp216.dm

```diff
@@ -0,0 +1,259 @@
+		/obj/item/clothing/under/syndicate/combat = 100,
+//		/obj/item/clothing/under/syndicate/terragov = 100,
+		/obj/item/clothing/suit/armor = 500,
+//		/obj/item/clothing/suit/armor/pcarrier/medium = 500,
+		/obj/item/clothing/suit/armor/riot = 450,
+		/obj/item/clothing/suit/armor/bulletproof = 450,
+		/obj/item/clothing/suit/armor/laserproof = 450,
+		/obj/item/clothing/suit/armor/vest/warden = 450,
+//		/obj/item/clothing/suit/armor/swat/officer = 400,
+//		/obj/item/clothing/suit/armor/vest/ert = 300,
+//		/obj/item/clothing/suit/armor/vest/ert/security = 300,
+//		/obj/item/clothing/suit/storage/vest/tactical = 250,
+//		/obj/item/clothing/suit/storage/vest/merc = 250,
+//		/obj/item/clothing/suit/armor/pcarrier/tan/tactical = 250,
+//		/obj/item/excalibur = 20,
+//		/obj/item/gun/projectile/pistol = 80,
+//		/obj/item/gun/projectile/pistol/military = 60,
+//		/obj/item/gun/projectile/automatic = 30,
+//		/obj/item/gun/projectile/heavysniper = 5,
+		/obj/item/market_uplink = 60,
+		/obj/item/multitool/uplink = 60,
+		/obj/item/soulstone = 50,
+		/obj/item/chameleon = 30,
+		/obj/item/assembly/flash/handheld = 20,
+//		/obj/item/reagent_containers/pill/scp500 = 10,
+//		/obj/item/storage/pill_bottle/scp500 = 1,
+		/mob/living/simple_animal/mouse/white = 150,
+		/mob/living/simple_animal/pet/dog/corgi = 100,
+		/mob/living/simple_animal/slime = 80,
+		/mob/living/simple_animal/hostile/carp = 50,
+		/mob/living/simple_animal/hostile/asteroid/wolf = 40,
+		/mob/living/simple_animal/hostile/asteroid/polarbear
+		)
+
+/obj/structure/scp216/Initialize()
+	. = ..()
+	SCP = new /datum/scp(
+		src, // Ref to actual SCP atom
+		"safe", //Name
+		SCP_SAFE, //Obj Class
+		"216", //Numerical Designation
+	)
+
+/obj/structure/scp216/Destroy()
+	LAZYCLEARLIST(all_codes) // Forever gone
+	return ..()
+
+/obj/structure/scp216/update_icon()
+	..()
+	if(open)
+		icon_state = "[initial(icon_state)]-open"
+	else
+		icon_state = initial(icon_state)
+
+/obj/structure/scp216/attackby(obj/item/I, mob/user)
+	if(!open)
+		return ..()
+	InsertItem(user, I, current_code)
+
+/obj/structure/scp216/attack_hand(mob/user)
+	if(!user.stat == CONSCIOUS)
+		to_chat(user, span_warning("You cannot use the safe while unconscious!"))
+		return
+	var/text_code = add_leading(num2text(current_code, 7), 7, "0")
+	var/dat = "<center>"
+	dat += "<a href='?src=\ref[src];open=1'>[open ? "Close" : "Open"] [src]</a><br>"
+	dat += "Current code: <a href='?src=\ref[src];change_code=1'>[text_code]</a><br>"
+	if(open && (num2text(current_code, 7) in all_codes))
+		dat += "<table>"
+		for(var/atom/movable/A in all_codes[num2text(current_code, 7)])
+			if(!ismob(A)) // Don't add mobs to the UI list
+				dat += "<tr><td><a href='?src=\ref[src];retrieve=\ref[A]'>[A.name]</a></td></tr>"
+		dat += "</table></center>"
+	var/datum/browser/popup = new(user, "safe", "Safe", 350, 300)
+	popup.set_content(dat)
+	popup.open()
+
+/obj/structure/scp216/Topic(href, href_list)
```

Comment:
This proc does not validate the user is actually nearby or able to interact with the structure, which leaves it open to href injection exploits. Try `usr.canUseTopic(src, USE_CLOSE|USE_NEED_HANDS|USE_IGNORE_TK)`